### ISSUE TITLE
Cross Compiling: failed to run custom build command

### ISSUE THREAD

**Issue by Fonsori** [briansmith/ring#1331]
1. This is the only issue remaining with compiling from m1 OSX to x86_64 Linux
2. Specifying linkers in .cargo/config does not affect anything
3. GNU or MUSL— not sure which to use but neither work. Both build commands fail, but because of different reasons
4. FiloSottile/musl-cross/musl-cross toolchain is in use 
5. Third day searching for a solution, looking grim 
6. This is my first gh issue— please don't hesitate to point out shortcomings 
7. Here's me trying to understand the full error:

cargo build --release --target=x86_64-unknown-linux-musl
   Compiling ring v0.16.20
   Compiling futures-core v0.3.15
   Compiling proc-macro-nested v0.1.7
   Compiling futures-task v0.3.15
   Compiling tracing v0.1.26
   Compiling http-body v0.4.2
   Compiling serde v1.0.126
   Compiling unicode-normalization v0.1.19
**error: failed to run custom build command for `ring v0.16.20`** 
>  _# ^ This line looks like the problem..._

Caused by:
  process didn't exit successfully: `/Users/skipper/Desktop/Rust/scout/target/release/build/ring-0c3d64667514526d/build-script-build` (exit status: 101)  
> _# ^ Seems like the build-script failed_

  --- stdout  _

> _# Does this config seem alright? Target as linux_gnu has a build command problem with ring as well._

  OPT_LEVEL = Some("3")
  TARGET = Some("x86_64-unknown-linux-musl")
  HOST = Some("aarch64-apple-darwin")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")

  --- stderr
  running "musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-I" "include" "-Wall" "-Wextra" "-pedantic" "-pedantic-errors" "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-fno-strict-aliasing" "-fvisibility=hidden" "-fstack-protector" "-g3" "-U_FORTIFY_SOURCE" "-DNDEBUG" "-c" "-o/Users/skipper/Desktop/Rust/scout/target/x86_64-unknown-linux-musl/release/build/ring-d1b1f8150d104d01/out/aesni-x86_64-elf.o" "/Users/skipper/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/aesni-x86_64-elf.S"

> _The gcc panic_

  thread 'main' panicked at 'failed to execute ["musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-I" "include" "-Wall" "-Wextra" "-pedantic" "-pedantic-errors" "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-fno-strict-aliasing" "-fvisibility=hidden" "-fstack-protector" "-g3" "-U_FORTIFY_SOURCE" "-DNDEBUG" "-c" "-o/Users/skipper/Desktop/Rust/scout/target/x86_64-unknown-linux-musl/release/build/ring-d1b1f8150d104d01/out/aesni-x86_64-elf.o" "/Users/skipper/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/aesni-x86_64-elf.S"]: 

> _Why does this directory does not exist here?_

No such file or directory (os error 2)', /Users/skipper/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/build.rs:653:9

**Comment by Fonsori** [comment]
Looking at more issues here, seems like cross compiling isn't working for yet more people, due to rustls as well.

**Comment by andrewtj** [comment]
The build script panic is due to being unable to run an executable called "musl-gcc". Have you tried setting [`TARGET_CC` and `TARGET_AR`](
https://github.com/briansmith/ring/blob/508f5a32934feb4ce5f9daac2bcfc978652dc3e0/BUILDING.md#cross-compiling)?

**Comment by Fonsori** [comment]
Yes, I've tried those— though it turns out [musl-cross](https://github.com/FiloSottile/homebrew-musl-cross) doesn't have M1 support. Embarrassingly, I've overlooked that. Using [macos-cross-toolchains](https://github.com/messense/homebrew-macos-cross-toolchains) compiled and ran on the target machine successfully, however. This is not an issue with Ring! 

Apologies for being mislead by the build error.
